salt/modules/vagrant.py

```diff
@@ -0,0 +1,532 @@
+    vm_['vagrant_provider'] = vagrant_provider or vm_.get('vagrant_provider', '')
+    _update_cache(name, vm_)
+
+    if start:
+        log.debug('Starting VM {0}'.format(name))
+        ret = _start(name, vm_)
+    else:
+        ret = 'Name {} defined using VM {}'.format(name, vm_['machine'] or '(default)')
+    return ret
+
+
+def _runas_sudo(vm_, command):
+    '''
+    prepend "sudo -u <runas> " if _vm['runas'] is defined
+    :param vm_: the virtual machine configuration dictionary
+    :param command: the command line which will be sent
+    :return: "sudo -u <runas> command" or "command" as needed
+    '''
+    runas = vm_.get('runas', False)
+    if runas:
+        return 'sudo -u {} {}'.format(runas, command)
+    return command
+
+
+def start(name):
+    '''
+    Start (vagrant up) a defined virtual machine by salt_id name.
+    The machine must have been previously defined using "vagrant.init".
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt <host> vagrant.start <salt_id>
+    '''
+    vm_ = get_vm_info(name)
+    return _start(name, vm_)
+
+
+def _start(name, vm_):  # internal call name, because "start" is a keyword argument to vagrant.init
+
+    try:
+        machine = vm_['machine']
+    except KeyError:
+        raise SaltInvocationError('No Vagrant machine defined for Salt-id {}'.format(name))
+
+    vagrant_provider = vm_.get('vagrant_provider', '')
+    provider_ = '--provider={}'.format(vagrant_provider) if vagrant_provider else ''
+    cmd = _runas_sudo(vm_, 'vagrant up {} {}'.format(machine, provider_))
+    log.info('Executing command "%s"', cmd)
+    ret = subprocess.call(
+            [cmd],
+            shell=True,
+            cwd=vm_.get('cwd', None)
+            )
+
+    return ret == 0
+
+
+def shutdown(name):
+    '''
+    Send a soft shutdown signal to the named vm.
+    ( for Vagrant, alternate name for vagrant.stop )
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt <host> vagrant.shutdown <salt_id>
+    '''
+    return stop(name)
+
+
+def stop(name):
+    '''
+    Hard shutdown the virtual machine.
+    ( Vagrant will attempt a soft shutdown first. )
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt <host> vagrant.stop <salt_id>
+    '''
+    vm_ = get_vm_info(name)
+    machine = vm_['machine']
+
+    cmd = _runas_sudo(vm_, 'vagrant halt {}'.format(machine))
+    log.info('Executing command "%s"', cmd)
+    ret = subprocess.call(
+        [cmd],
+        shell=True,
+        cwd=vm_.get('cwd', None)
+        )
+    return ret
+
+
+def pause(name):
+    '''
+    Pause (suspend) the named vm
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt <host> vagrant.pause <salt_id>
+    '''
+    vm_ = get_vm_info(name)
+    machine = vm_['machine']
+
+    cmd = _runas_sudo(vm_, 'vagrant suspend {}'.format(machine))
+    log.info('Executing command "%s"', cmd)
+    ret = subprocess.call(
```

Comment:
Is there a reason to use `subprocess.call` instead of `__salt__['cmd.shell']`?

The other benefit is that in 2017.7, runas is a state runtime variable, so that you don't have to provide runas in the state if you write a vargant state as well, the cmd.shell will just get the runas user from the `__context__` object.

And then you also do not need the _runas_sudo command, you can just pass which user to runas to the cmd module.

Reply:
I did not even think about using __salt__['cmd.shell'].  Too used to doing it all myself. I'll look at that.

Yes, I am planning to write a vagrant state module, just need this as a starting point.

Reply:
Great suggestion!  The module is now 70 lines shorter.